V0.2

dmcharlie19 opened the pull request.

dmcharlie19 committed:
компонент регистрации

сделана форма регистрации и валидация параметров

dmcharlie19 committed:
форма аутентификации

сделана форма аутентификации и валидация параметров

dmcharlie19 committed:
аутентификаия

сервис аутентификации на стороне клиента
конечная точка на сервере
перехватчик ошибок
сервис ошибок

dmcharlie19 committed:
переименование неймспейса под общий стиль

dmcharlie19 committed:
аутентификация на стороне клиента

Реализовано получение жвт токена, его сохранение и отправка на сервер через интерцептор

dmcharlie19 committed:
Аутентификация и регистрация на стороне сервера

создана таблица пользователей
апи регистрации
поддержка регистрации на клиенте
доработка сервиса ошибок и отображение ошибок коммуникации

egoshin-igor commented on CookBookBackend/CookBook.Api/Controllers/AccountController.cs:
```diff
@@ -0,0 +1,83 @@
+    [HttpPost, Route( "login" )]
+    public IActionResult Authentication( [FromBody] AuthenticationRequestDto authRequestDto )
+    {
+      if ( authRequestDto.Login == null )
```
Создашь AcoountService в Api слое? С методами Registrate, Authenticate

dmcharlie19 replied:
сделал

egoshin-igor commented on CookBookBackend/CookBook.Application/Entities/Users/AuthOptions.cs:
```diff
@@ -0,0 +1,18 @@
+{
+  public class AuthOptions
+  {
+    private const string KEY = "CookBookClientCookBookServer";
```
Константы в CamelCase -> Key, Issuer и т.п.

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookFrontend/src/app/interceptors/error-interceptor.ts:
```diff
@@ -0,0 +1,43 @@
+        console.log("intercept!");
+        this.errorService.clearError();
+
+        var idToken = this.accountService.getAccesToken();
```
var не юзаем в ts. const если не меняется переменная, let если меняется ниже

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookFrontend/src/app/interceptors/error-interceptor.ts:
```diff
@@ -0,0 +1,43 @@
+        console.log("intercept!");
+        this.errorService.clearError();
+
+        var idToken = this.accountService.getAccesToken();
```
типизируем все типы -> const idToken: string

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookFrontend/src/app/interceptors/error-interceptor.ts:
```diff
@@ -0,0 +1,43 @@
+        console.log("intercept!");
+        this.errorService.clearError();
+
+        var idToken = this.accountService.getAccesToken();
```
Именуем правильно. В данном случае это accessToken

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookFrontend/src/app/recipes-list/recipes-list.component.html:
```diff
@@ -49,12 +51,14 @@
         <app-recipe-card [recipe]=recipe></app-recipe-card>
     </div>
-    <div *ngIf="isLoaded">
+    <!-- <div *ngIf="isLoaded">
```
комментарии если они не нужно удаляй

dmcharlie19 replied:
убрал

egoshin-igor commented on CookBookFrontend/src/app/services/accountService.ts:
```diff
@@ -0,0 +1,78 @@
+        return this.AccesToken == null;
+    }
+
+    // getExpiration() {
```
лишний коммент

dmcharlie19 committed:
компонент добавления рецепта

верстка + валидация

dmcharlie19 committed:
добавление поля количество персон в модель рецепта

dmcharlie19 committed:
Добавление сущностей RecipeStep RecipeIngridient

добавлены модели
репоситории
конфигурации
миграции БД

dmcharlie19 committed:
переименование неймспейса

dmcharlie19 committed:
добавление сущности Tag

добавлены домены Tag и TagRecipes
реализовано добавление тегов в БД и связывание их с рецептами

dmcharlie19 committed:
доработка формы добавления рецепта и запроса списка рецептов

dmcharlie19 committed:
доработки по стилю

dmcharlie19 committed:
перенос бизнес логики из контроллеров в сервисы

+реализован перехватчик исключений

dmcharlie19 committed:
Исправления по замечаниям

egoshin-igor commented on CookBookBackend/CookBook.Api/Controllers/RecipesController.cs:
```diff
@@ -1,25 +1,43 @@
     }
 
-    [HttpGet( "" )]
+    [HttpGet, Route( "" )]
```
Пустой роут можно не писать. Просто HttpGet

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookBackend/CookBook.Api/Controllers/RecipesController.cs:
```diff
@@ -1,25 +1,43 @@
     }
+
+    [HttpPost, Authorize, Route( "AddRecipe" )]
+    public void AddRecipe( [FromBody] AddRecipeRequestDto addRecipeRequest )
```
UserId откуда берешь? Из AddRecipeRequestDto  если, то неправильно. Через апи тогда можно будет добавить рецепт за другого юзера. Бери userId из Claims

dmcharlie19 replied:
я еще не сделал связь рецепта и пользователя

dmcharlie19 replied:
сделал

egoshin-igor commented on CookBookBackend/CookBook.Api/Migrations/20220513135632_AddFieldPersonCountInRecipe.cs:
```diff
@@ -0,0 +1,26 @@
+      migrationBuilder.AddColumn<int>(
+          name: "PersonCount",
+          table: "Recipes",
+          type: "int",
```
tinyint можно какой нибудь юзать(byte в C#) но это так написал. Можешь не делать

dmcharlie19 replied:
ладно

egoshin-igor commented on CookBookBackend/CookBook.Application/Entities/Users/AuthOptions.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace CookBook.Application.Entities.Users
+{
+  public class AuthOptions
```
Проверь такой кейс у JWT - авторизовался, access token устарел, клиент получает 403, ui должен будет рефрешнуть access token по рефреш токену и еще раз тот же запрос переотправляет с новым токеном. Если этого нет, то не все сделано по JWT.

dmcharlie19 replied:
рефреш токен не делал еще

dmcharlie19 replied:
сделал выход по истечению таймаута

egoshin-igor commented on CookBookBackend/CookBook.Application/Exceptions/InvalidClientParameterException.cs:
```diff
@@ -0,0 +1,15 @@
+    public InvalidClientParameterException( string message ) : base( message ) { }
+    public InvalidClientParameterException( string message, Exception inner ) : base( message, inner ) { }
+
+    protected InvalidClientParameterException( System.Runtime.Serialization.SerializationInfo info,
```
Зачем тебе этот конструктор? Он нужен точно?

dmcharlie19 replied:
нет, убрал

egoshin-igor commented on CookBookBackend/CookBook.Application/Repositories/IRecipeIngredientRepository.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace CookBook.Application.Repositories
+{
+  public interface IRecipeIngredientRepository
```
Этого репозитория не будет. У тебя есть рецепт, а внутри ингредиенты. Когда в репу добавляешь рецепт, то ингредиенты тоже добавляются

dmcharlie19 replied:
убрал

egoshin-igor commented on CookBookBackend/CookBook.Application/Repositories/IRecipeStepRepository.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace CookBook.Application.Repositories
+{
+  public interface IRecipeStepRepository
```
Аналогично IRecipeIngredientRepository. Нужно убрать

dmcharlie19 replied:
убрал

egoshin-igor commented on CookBookBackend/CookBook.Application/Repositories/ITagRecipeRepository.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace CookBook.Application.Repositories
+{
+  public interface ITagRecipeRepository
```
Аналогично IRecipeIngredientRepository. Нужно убрать

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookBackend/CookBook.Application/Repositories/IUserRepository.cs:
```diff
@@ -0,0 +1,12 @@
+  public interface IUserRepository
+  {
+    void Add( User user );
+    void Remove( User user );
```
Этот метод не нужен? Есть кейс удаления юзера?

dmcharlie19 replied:
пока нет, убрал

egoshin-igor commented on CookBookBackend/CookBook.Application/Services/AccountService.cs:
```diff
@@ -0,0 +1,74 @@
+        registarationDto.Password,
+        registarationDto.Name ) );
+
+      _unitOfWork.Commit();
```
коммит на уровень выше - к контроллеру

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookBackend/CookBook.Application/Services/Dto/AuthenticationResponseDto.cs:
```diff
@@ -0,0 +1,9 @@
+  {
+    public string? AccesToken { get; set; }
+    public string? UserName { get; set; }
+    public int ExpiresAt { get; set; }
```
DateTime тут должен быть

dmcharlie19 replied:
исправил, также сделал разлогирование по истечению времени

egoshin-igor commented on CookBookBackend/CookBook.Application/Services/RecipeService.cs:
```diff
@@ -0,0 +1,98 @@
+        likesCount: 0,
+        favoritesCount: 0 );
+
+      recipe.RecipeSteps = addRecipeRequest.CookingSteps
```
recipe.RecipeSteps = addRecipeRequest.CookingSteps
        .Select( recipeStep => new RecipeStep( recipeStep, recipe.Id ) )
        .ToList();

      recipe.RecipeIngredients = addRecipeRequest.RecipeIngridients
        .Select( ri => new RecipeIngredient( ri.IngridientTitle, ri.IngridientBody, recipe.Id ) )
        .ToList();
можно это в конструктор вынести или сделать у рецепта методы - AddSteps, AddIngredients, Там можно и валидацию модели сделать внутри метода в Recipe

dmcharlie19 replied:
ок сделал

egoshin-igor commented on CookBookBackend/CookBook.Application/Services/RecipeService.cs:
```diff
@@ -0,0 +1,98 @@
+      // #TODO Валидация модели
+
+      _recipeRepository.Add( recipe );
+      _unitOfWork.Commit();
```
смотри что предлагаю - выносим uow выше в контроллер и делаем такую логику
Сервис тегов
-- сперва добавляем теги
-- возвращаем List<Tag>(созданные или уже полученные с базы)
Uow.Commit
Новые теги сохраняются у них в Tag автоматом Id появится(фреймворк проставит)
Собираем команду добавления рецепта. Поле List<Tags> будет в команде уже с тегами с Id(просто список тегов с Id у каждого тэга)
Сервис рецептов
-- Добавляем рецепт. Достаем TagId так
----список тегов в словарь - Tags.ToDictionary(x => x.Name) = tagByName
----RecipeTags получаем так command.RecipeTags.Select(x => { tag = tagByName[x.Name]; return new RecipeTag(  tag.Id ) Ну и валидация что тег есть в словаре, если нет, то исключение(такого не должно быть)
Uow.Commit в контроллере

egoshin-igor replied:
Есть еще такая штука --- SequenceHillo называется у mssql. При добавлении в репозитории, но до SaveChanges бд генерит Id заранее для сущности и отдает сразу приложению. Что это дает - у сущностей уже есть Id до сохранения сущности в таблицу. Это дает возможность операции сохранения и тэга и рецепта в одной Uow.Commit( в 1 транзакции). Удобно. Это если вдруг захочешь заюзать.

egoshin-igor replied:
ef core тоже поддерживает SequenceHilo

dmcharlie19 replied:
Вроде все сделал, но с SequenceHilo разобраться не смог, хотя выглядит интересно

egoshin-igor commented on CookBookBackend/CookBook.Application/Services/RecipeService.cs:
```diff
@@ -0,0 +1,98 @@
+        .ToList();
+
+      recipe.RecipeIngredients = addRecipeRequest.RecipeIngridients
+        .Select( ri => new RecipeIngredient( ri.IngridientTitle, ri.IngridientBody, recipe.Id ) )
```
recipe.Id   не передавать. Т.к. это дочерняя сущность, то фреймворк сам проставит его

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookBackend/CookBook.Infrastructure/Repositories/RecipeIngredientRepository.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace CookBook.Infrastructure.Repositories
+{
+  public class RecipeIngredientRepository : IRecipeIngredientRepository
```
это все уберется

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookFrontend/src/app/interceptors/error-interceptor.ts:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+        console.log("intercept!");
```
консоль логи - это дебаг. В коде его не должно быть. Только во время дебага

dmcharlie19 replied:
убрал

egoshin-igor commented on CookBookFrontend/src/app/services/accountService.ts:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    public logout() {
+
```
пустая стркоа

dmcharlie19 replied:
убрал

egoshin-igor commented on CookBookFrontend/src/app/services/errorService.ts:
```diff
@@ -0,0 +1,30 @@
+
+    public handleError(error: HttpErrorResponse): void {
+
+        var errorMessage: string = "";
```
var на let

dmcharlie19 replied:
ок

dmcharlie19 committed:
перенос хранения токенов из класса в локальное хранилище браузера

dmcharlie19 committed:
валидация модели рецепта

и исправление небольших недочетов
удалены лишние репозитории

dmcharlie19 committed:
Добавление сервиса тэгов

dmcharlie19 committed:
Добавление валидации во все модели

dmcharlie19 committed:
заготовка компонента деталей рецепта

dmcharlie19 committed:
добавлена связь пользователя и рецепта

отображение автора рецепта
передача на фронт авторов
миграция
доработана модель

dmcharlie19 committed:
реализация компенента с деталями рецепта

dmcharlie19 committed:
запрос полной информации о рецепте

dmcharlie19 committed:
доработка деталей рецепта

dmcharlie19 committed:
исправления по замечаниям. разлогирование при истечении таймаута

egoshin-igor commented on CookBookBackend/CookBook.Core/Domain/Recipe.cs:
```diff
@@ -1,29 +1,113 @@
-      FavoritesCount = favoritesCount;
+        public int Id { get; protected set; }
+
+        [Required]
```
можешь оставить конечно. Но такую валидацию в модель не суют обычно. Слишком низкоуровневые вещи. Обычно так валидируют команду создания рецепта, или логику проверки в обновлении такую делают.

dmcharlie19 replied:
ладно

egoshin-igor commented on CookBookBackend/CookBook.Core/Domain/Recipe.cs:
```diff
@@ -1,29 +1,113 @@
+                throw new InvalidClientParameterException( "���������� ������� �� ������ ���� �������" );
+
+            if ( ingredients.Count >= _maxIngredientsCount )
+                throw new InvalidClientParameterException( "��������� ������������ ���������� ������������" );
```
кодировку поправь

dmcharlie19 replied:
ок

egoshin-igor commented on CookBookBackend/CookBook.Core/Domain/Recipe.cs:
```diff
@@ -1,29 +1,113 @@
+            if ( tags.Count >= _maxTagsCount )
+                throw new InvalidClientParameterException( "��������� ������������ ���������� �����" );
+
+            Tags = new();
```
у тебя метод врет. Написано AddTags, а он полностью перезаписывает вместо добавления. Название мб поменяй

dmcharlie19 replied:
ок, поменял логику

dmcharlie19 committed:
доработки по ревью
